### PR TITLE
Switch to packaged Thrust on Ubuntu, enable CentOS 7.5 as a CI target

### PR DESCRIPTION
**Pull request by iotamudelta**
1) Use the hip-thrust version of Thrust as opposed to the GH master. (ROCm 267)

2) CentOS 7.5 docker (ROCm 279)

* Always install the libraries at docker creation for ubuntu.
* Add Dockerfile for CentOS ROCm
* Enable the centos build
* Source devtoolset in bashrc
* Set locales correctly depending on whether we are on Ubuntu or CentOS
* Install a newer cmake for CentOS
* Checkout thrust as there is no package for CentOS yet.

PyTorch/Caffe2 on ROCm passed tests: https://github.com/ROCmSoftwarePlatform/pytorch/pull/280

For attention: @bddppq @ezyang 

Docker rebuild for Ubuntu not urgent (getting rid of Thrust checkout and package install is mainly cosmetic). If docker for CentOS 7.5 is wanted, build is necessary. Build of PyTorch tested by me in CentOS docker. PyTorch unit tests work mostly, however, a test in test_jit causes a python recursion error that seems to be due to the python2 on CentOS as we haven't ever seen this on Ubuntu - hence please do not enable unit tests.

### CODE REVIEW

File: .jenkins/pytorch/build.sh [hunk 1]
```diff
@@ -43,22 +43,24 @@ cmake --version
 pip install -r requirements.txt || true
 
 if [[ "$BUILD_ENVIRONMENT" == *rocm* ]]; then
+
+  if [[ "$BUILD_ENVIRONMENT" == *centos* ]]; then
```
Comment (bddppq) [inline .jenkins/pytorch/build.sh]: Since we are going to build new docker images with this PR, let's export these in the docker file.

File: caffe2/core/hip/context_hip.h [hunk 2]
```diff
@@ -175,7 +175,7 @@ class HIPContext final : public BaseContext {
     return hip_objects_.GetStream(gpu_id, stream_id);
   }
 
-  rocblas_handle rocblas_handle() {
+  rocblas_handle rocblashandle() {
```
Comment (bddppq) [inline caffe2/core/hip/context_hip.h]: looks like this fix is unrelated to docker image? since the docker PR will takes longer to land (need to rebuild docker images and do testing and update build config) could you put this fix in a separate PR so we can land it sooner?

File: cmake/public/LoadHIP.cmake [hunk 3]
```diff
@@ -62,11 +62,8 @@ ENDIF()
 # THRUST_PATH
 IF(DEFINED ENV{THRUST_PATH})
   SET(THRUST_PATH $ENV{THRUST_PATH})
-ELSEIF(DEFINED ENV{THRUST_ROOT})
-  # TODO: Remove support of THRUST_ROOT environment variable
-  SET(THRUST_PATH $ENV{THRUST_ROOT})
 ELSE()
-  SET(THRUST_PATH ${ROCM_PATH}/Thrust)
+  SET(THRUST_PATH ${ROCM_PATH}/include)
```
Comment (bddppq) [inline cmake/public/LoadHIP.cmake]: Are you sure this is correct? ("include" seems to be the header include path, not the root path)

File: docker/caffe2/jenkins/centos-rocm/.gitignore [hunk 4]
```diff
@@ -0,0 +1 @@
+*.sh
```
Comment (bddppq) [inline docker/caffe2/jenkins/centos-rocm/.gitignore]: this gitingore file doesn't seem needed

File: docker/caffe2/jenkins/common/install_rocm.sh [hunk 5]
```diff
@@ -42,19 +46,55 @@ install_ubuntu() {
...
 install_centos() {
-    echo "Not implemented yet"
-    exit 1
+
+  yum update -y
+  yum install -y wget
+  yum install -y openblas-devel
+
+  yum install -y centos-release-scl
+  yum install -y devtoolset-7
```
Comment (bddppq) [inline docker/caffe2/jenkins/common/install_rocm.sh]: can we put this to a separate install_devtoolset.sh? and then add "-devtoolset7-" in our docker image name.

**Review by bddppq — COMMENTED** [review]
Nice work!

**Comment by bddppq** [comment]
Does devtoolset-7 have llvm-7?

File: cmake/public/LoadHIP.cmake (hunk 3 above)
Comment (iotamudelta) [reply cmake/public/LoadHIP.cmake]: Yes, that's correct. The hip-thrust package is putting headers into `/opt/rocm/include/thrust/` so in order to access them with `#include <thrust/something.h`, we need that.

File: docker/caffe2/jenkins/centos-rocm/.gitignore (hunk 4 above)
Comment (iotamudelta) [reply docker/caffe2/jenkins/centos-rocm/.gitignore]: It comes with the others. There is some soft-linking of shell scripts going on during the build, so the gitignore for them makes sense to me?

File: caffe2/core/hip/context_hip.h (hunk 2 above)
Comment (iotamudelta) [reply caffe2/core/hip/context_hip.h]: It's somewhat related as that the docker image uses `gcc` as the base compiler which actually complains about this. `llvm` tolerates it. I can make a separate PR for it though.

**Comment by iotamudelta** [comment]
No, devtoolset-7 brings a newer gcc and related packages. It's a requirement for ROCm on CentOS.

File: .jenkins/pytorch/build.sh (hunk 1 above)
Comment (iotamudelta) [reply .jenkins/pytorch/build.sh]: Which file is that?

File: docker/caffe2/jenkins/common/install_rocm.sh (hunk 5 above)
Comment (iotamudelta) [reply docker/caffe2/jenkins/common/install_rocm.sh]: Sure.

File: .jenkins/pytorch/build.sh (hunk 1 above)
Comment (bddppq) [reply .jenkins/pytorch/build.sh]: the new docker/caffe2/jenkins/centos-rocm/Dockerfile (and also the ubuntu-rocm/Dockerfile

File: docker/caffe2/jenkins/centos-rocm/.gitignore (hunk 4 above)
Comment (bddppq) [reply docker/caffe2/jenkins/centos-rocm/.gitignore]: oh you are right, I didn't notice other directories have this as well

**Comment by bddppq** [comment]
@iotamudelta Doesn't hcc need to use clang-7 as the base compiler? (as in the ubuntu docker image)

File: docker/caffe2/jenkins/build.sh [hunk 6]
```diff
@@ -39,6 +39,10 @@ fi
 if [[ "$image" == *rocm* ]]; then
   ROCM_VERSION="$(echo "${image}" | perl -n -e'/rocm(\d+\.\d+\.\d+|nightly)/ && print $1')"
   DOCKERFILE="${OS}-rocm/Dockerfile"
+  if [[ "$image" == *centos* ]]; then
```
Comment (bddppq) [inline docker/caffe2/jenkins/build.sh]: let's remove this "if" (to make it the same for ubuntu as well)

File: .jenkins/pytorch/build.sh (hunk 1 above)
Comment (iotamudelta) [reply .jenkins/pytorch/build.sh]: They should be exported already in both of them, see lines 45/46 in `docker/caffe2/jenkins/centos-rocm/Dockerfile` and lines 10/11 in `docker/caffe2/jenkins/ubuntu-rocm/Dockerfile `

File: docker/caffe2/jenkins/common/install_rocm.sh [hunk 7]
```diff
@@ -41,20 +45,60 @@ install_ubuntu() {
     sed -i 's/find_dependency(hip)/find_dependency(HIP)/g' /opt/rocm/rocblas/lib/cmake/rocblas/rocblas-config.cmake
 }
 
+install_devtoolset() {
```
Comment (bddppq) [inline docker/caffe2/jenkins/common/install_rocm.sh]: I meant putting it in a separate shell script (so other non-rocm docker images can use it to control devtoolset as well)

Comment (iotamudelta) [reply docker/caffe2/jenkins/common/install_rocm.sh]: OK.

**Comment by iotamudelta** [comment]
@bddppq no, `hcc` does not require LLVM as the base compiler. Issues just arise when you cross ABI between base compiler and `hcc` (such as passing objects by reference between code objects compiled w/ different compilers).

File: .jenkins/pytorch/build.sh (hunk 1 above)
Comment (bddppq) [reply .jenkins/pytorch/build.sh]: sorry I meant LC_ALL and LANG environment variables, not BUILD_ENVIRONMENT

File: docker/caffe2/jenkins/common/install_devtoolset.sh [hunk 8]
```diff
@@ -0,0 +1,10 @@
+#!/bin/bash
+
+set -ex
+
+[ -n "$CENTOS_VERSION" ]
+
+yum install -y centos-release-scl
+yum install -y devtoolset-$CENTOS_VERSION
```
Comment (bddppq) [inline docker/caffe2/jenkins/common/install_devtoolset.sh]: wait I think CENTOS_VERSION and devtoolset version are decoupled.

Comment (iotamudelta) [reply docker/caffe2/jenkins/common/install_devtoolset.sh]: Yes, somewhat but let me do this more proper.

File: caffe2/core/hip/context_hip.h (hunk 2 above)
Comment (bddppq) [reply caffe2/core/hip/context_hip.h]: @iotamudelta ping to split out this into separate PR

Comment (iotamudelta) [reply caffe2/core/hip/context_hip.h]: @bddppq The PR is in preparation. Should come through CI hopefully soon.

Comment (bddppq) [reply caffe2/core/hip/context_hip.h]: @iotamudelta could you remove the rocblas_handle changes from this PR?

**Comment by bddppq** [comment]
triggered rocm CI with new docker image:

https://ci.pytorch.org/jenkins/job/caffe2-builds/job/py2-clang7-rocmdeb-ubuntu16.04-trigger-test/2818/
https://ci.pytorch.org/jenkins/job/pytorch-builds/job/py2-clang7-rocmdeb-ubuntu16.04-trigger/2817/

**Comment by iotamudelta** [comment]
@bddppq if you wanted to switch to 1.9.2 prerelease here, you'd need `6441786` (see: https://github.com/ROCmSoftwarePlatform/pytorch/pull/290 ). Note however that this still wouldn't link, so I am not sure we want to do that?

If you do want to switch, I can cherrypick that commit in here (or you can obviously also do it).

**Comment by bddppq** [comment]
triggered builds with new docker image

caffe2 ubuntu: https://ci.pytorch.org/jenkins/job/caffe2-builds/job/py2-clang7-rocmdeb-ubuntu16.04-trigger-test/6037
pytorch ubuntu: https://ci.pytorch.org/jenkins/job/pytorch-builds/job/py2-clang7-rocmdeb-ubuntu16.04-trigger/6026/
caffe2 centos: https://ci.pytorch.org/jenkins/job/caffe2-builds/job/py2-devtoolset7-rocmrpm-centos7.5-trigger-test/20/
pytorch centos: https://ci.pytorch.org/jenkins/job/pytorch-builds/job/py2-devtoolset7-rocmrpm-centos7.5-trigger/9/

**Review by bddppq — APPROVED** [review]
